From a17be29c6b6cbbb6e4d07ec357f924aacbd4307f Mon Sep 17 00:00:00 2001
From: Christophe Kerello <christophe.kerello@foss.st.com>
Date: Tue, 18 Mar 2025 09:28:46 +0100
Subject: [PATCH 810/841] feat(stm32mp2): pwr/syscfg: allow to set a fixed IO
 compensation

Based on new property st,iocomp, allow to set a fixed IO compensation
or to use the dynamic IO compensation.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git 3b07ebdff17a86d46d2ae32cf143e737c2154713 feat(stm32mp2): pwr/syscfg: allow to set a fixed IO compensation]
Change-Id: I177137af7c6b20f9f47f12c474b2673404e85b87
Signed-off-by: Christophe Kerello <christophe.kerello@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/444755
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Maxime MERE <maxime.mere@foss.st.com>
Tested-by: Christophe KERELLO <christophe.kerello@st.com>
Reviewed-by: Christophe KERELLO <christophe.kerello@st.com>
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
---
 plat/st/stm32mp2/include/stm32mp2_private.h |  3 ++-
 plat/st/stm32mp2/stm32mp2_pwr.c             | 20 +++++++++++++++++--
 plat/st/stm32mp2/stm32mp2_syscfg.c          | 22 ++++++++++++++++++++-
 3 files changed, 41 insertions(+), 4 deletions(-)

diff --git a/plat/st/stm32mp2/include/stm32mp2_private.h b/plat/st/stm32mp2/include/stm32mp2_private.h
index d13282d01..840704d68 100644
--- a/plat/st/stm32mp2/include/stm32mp2_private.h
+++ b/plat/st/stm32mp2/include/stm32mp2_private.h
@@ -33,7 +33,8 @@ enum syscfg_io_ids {
 	SYSFG_NB_IO_ID
 };
 
-void stm32mp_syscfg_enable_io_compensation(enum syscfg_io_ids id);
+void stm32mp_syscfg_enable_io_comp(enum syscfg_io_ids id);
+void stm32mp_syscfg_fixed_io_comp(enum syscfg_io_ids id, uint32_t pmos, uint32_t nmos);
 
 void stm32mp_syscfg_set_icn_qos(void);
 
diff --git a/plat/st/stm32mp2/stm32mp2_pwr.c b/plat/st/stm32mp2/stm32mp2_pwr.c
index 514a82aa6..6eb48da15 100644
--- a/plat/st/stm32mp2/stm32mp2_pwr.c
+++ b/plat/st/stm32mp2/stm32mp2_pwr.c
@@ -5,6 +5,7 @@
  */
 
 #include <common/debug.h>
+#include <common/fdt_wrappers.h>
 #include <drivers/st/regulator.h>
 #include <lib/mmio.h>
 #include <lib/utils_def.h>
@@ -14,13 +15,14 @@
 #include <stm32mp2_private.h>
 
 #define IO_VOLTAGE_THRESHOLD_MV 2700
+#define IOCOMP_CODE_MAX 2
 
 struct pwr_regu {
 	const char *node_name;
 	uint32_t enable_reg;
 	uint32_t valid_mask;
 	uint32_t vrsel_mask;
-	int comp_idx;
+	enum syscfg_io_ids comp_idx;
 	const char *supply_name;
 };
 
@@ -166,6 +168,7 @@ int stm32mp2_pwr_init_io_domains(void)
 		}
 
 		if (stm32mp2_pwr_domain_is_enabled(regu)) {
+			uint32_t iocomp_code[IOCOMP_CODE_MAX] = {0U};
 			int ret;
 
 			VERBOSE("Init IO domain %d\n", i);
@@ -175,7 +178,20 @@ int stm32mp2_pwr_init_io_domains(void)
 				return ret;
 			}
 
-			stm32mp_syscfg_enable_io_compensation(regu->comp_idx);
+			ret = fdt_read_uint32_array(fdt, subnode, "st,iocomp",
+						    IOCOMP_CODE_MAX,
+						    iocomp_code);
+			if ((ret != 0) && (ret != -FDT_ERR_NOTFOUND)) {
+				return ret;
+			}
+
+			if (ret == -FDT_ERR_NOTFOUND) {
+				stm32mp_syscfg_enable_io_comp(regu->comp_idx);
+			} else {
+				stm32mp_syscfg_fixed_io_comp(regu->comp_idx,
+							     iocomp_code[0],
+							     iocomp_code[1]);
+			}
 		}
 	}
 
diff --git a/plat/st/stm32mp2/stm32mp2_syscfg.c b/plat/st/stm32mp2/stm32mp2_syscfg.c
index 870a06a2c..160fbdd5f 100644
--- a/plat/st/stm32mp2/stm32mp2_syscfg.c
+++ b/plat/st/stm32mp2/stm32mp2_syscfg.c
@@ -47,6 +47,7 @@
 #define SYSCFG_CCCR_CS			BIT(9)
 #define SYSCFG_CCCR_EN			BIT(8)
 #define SYSCFG_CCCR_RAPSRC_MASK		GENMASK(7, 4)
+#define SYSCFG_CCCR_RAPSRC_SHIFT	4U
 #define SYSCFG_CCCR_RANSRC_MASK		GENMASK(3, 0)
 
 /* IO compensation CCSR registers bit definition */
@@ -162,7 +163,7 @@ size_t stm32mp_syscfg_get_mm_size(uint8_t bank)
  * @param  id: IO compensation ID
  * @retval 0.
  */
-void stm32mp_syscfg_enable_io_compensation(enum syscfg_io_ids id)
+void stm32mp_syscfg_enable_io_comp(enum syscfg_io_ids id)
 {
 	uintptr_t cccr_addr = SYSCFG_BASE + syscfg_cccr_offset[id];
 	uintptr_t ccsr_addr = cccr_addr + 4U;
@@ -188,6 +189,25 @@ void stm32mp_syscfg_enable_io_compensation(enum syscfg_io_ids id)
 	mmio_clrbits_32(cccr_addr, SYSCFG_CCCR_CS);
 }
 
+/*
+ * @brief  Set fixed IO compensation for an IO domain.
+ * @param  id: IO compensation ID
+ * @param  pmos: pmos compensation code
+ * @param  nmos: nmos compensation code
+ * @retval 0.
+ */
+void stm32mp_syscfg_fixed_io_comp(enum syscfg_io_ids id, uint32_t pmos, uint32_t nmos)
+{
+	uintptr_t cccr_addr = SYSCFG_BASE + syscfg_cccr_offset[id];
+	uint32_t value = 0U;
+
+	VERBOSE("Set fixed IO comp for id %u\n", id);
+
+	value = ((pmos << SYSCFG_CCCR_RAPSRC_SHIFT) & SYSCFG_CCCR_RAPSRC_MASK) |
+		(nmos & SYSCFG_CCCR_RANSRC_MASK);
+	mmio_write_32(cccr_addr, value);
+}
+
 #if STM32MP_SPI_NAND || STM32MP_SPI_NOR || STM32MP_HYPERFLASH
 /*
  * SYSCFG DLYBOS register offsets (base relative)
-- 
2.47.2

